File: packages/cli/src/commands/discover.ts
// cli/src/commands/discover.ts

import type { Command } from 'commander';
import ora from 'ora';
import chalk from 'chalk';
import inquirer from 'inquirer';
import { parseComponentIdCsv } from '../csv_parser.js';
import { promises as fs } from 'fs';
import { initiateDiscovery, pollForPlanCompletion, PlanFailedError } from '../api_client.js';
import type { CliPlanComponent } from '../types.js';
import { handleCliError } from '../error_handler.js';

export function registerDiscoverCommand(program: Command) {
  program
    .command('discover')
    .description('Create a new test plan from a list of components.')
    .option('--from-csv <path>', 'Path to a CSV file with a single column of component IDs.')
    .option('--dependencies', 'Discover all dependencies for the provided components.', false)
    .requiredOption('--creds <profile>', 'The name of the credential profile to use')
    .action(async (options) => {
      const spinner = ora('Preparing test plan...').start();
      try {
        const { fromCsv, dependencies, creds: profileName } = options;
        let componentIds: string[] = [];

        if (fromCsv) {
          spinner.text = `Reading components from ${chalk.cyan(fromCsv)}...`;
          try {
            const fileContent = await fs.readFile(fromCsv, 'utf-8');
            componentIds = parseComponentIdCsv(fileContent);
          } catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            throw new Error(`Failed to read or parse the CSV file at ${fromCsv}: ${message}`);
          }
        } else {
          spinner.stop(); // Stop spinner for interactive prompt
          const responses = [];
          let answer;
          const message = dependencies
            ? 'Enter a root Component ID to discover dependencies from (leave blank to finish):'
            : 'Enter a Component ID to add to the plan (leave blank to finish):';

          do {
            answer = await inquirer.prompt([
              {
                type: 'input',
                name: 'componentId',
                message: message
              }
            ]);
            if (answer.componentId) responses.push(answer.componentId);
          } while (answer.componentId);
          componentIds = responses;
          spinner.start();
        }

        if (componentIds.length === 0) {
          spinner.warn('No component IDs provided. Exiting.');
          return;
        }

        const discoveryMode = dependencies ? 'and all their dependencies' : '...';
        spinner.text = `Creating test plan with ${chalk.bold(componentIds.length)} component(s) ${discoveryMode}`;

        const { planId } = await initiateDiscovery(componentIds, profileName, dependencies);
        spinner.text = `Test plan created with ID: ${chalk.cyan(planId)}. Waiting for component processing to complete...`;

        const finalPlan = await pollForPlanCompletion(planId);
        spinner.succeed(chalk.green('Test plan processing complete!'));

        console.log(`\nTest Plan ID: ${chalk.cyan(planId)}`);

        // Renders a table that correctly displays the one-to-many relationship
        const displayData = finalPlan.planComponents.flatMap((comp: CliPlanComponent) => {
          if (comp.availableTests.length === 0) {
            return [{
              'Component ID': comp.componentId,
              'Component Name': comp.componentName || 'N/A',
              'Has Test Coverage': '❌ No',
              'Available Test ID': 'N/A',
            }];
          }
          return comp.availableTests.map(testId => ({
            'Component ID': comp.componentId,
            'Component Name': comp.componentName || 'N/A',
            'Has Test Coverage': '✅ Yes',
            'Available Test ID': testId,
          }));
        });

        if (displayData.length > 0) {
          console.table(displayData);
        } else {
          console.log(chalk.yellow('No components were found for this test plan.'));
        }

        console.log(chalk.yellow(`\nTo execute tests, use the 'execute' command with the Plan ID and desired Test IDs.`));

      } catch (error: any) {
        spinner.fail(chalk.red('Test plan creation failed.'));
        handleCliError(error);
      }
    });
}

File: packages/cli-go/cmd/discover.go
// cli-go/cmd/discover.go
package cmd

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/automated-test-orchestrator/cli-go/internal/client"
	"github.com/automated-test-orchestrator/cli-go/internal/csv"
	"github.com/automated-test-orchestrator/cli-go/internal/display"
	"github.com/briandowns/spinner"
	"github.com/fatih/color"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

// discoverCmd represents the discover command
var discoverCmd = &cobra.Command{
	Use:   "discover [componentIds...]",
	Short: "Create a new test plan from a list of components",
	Long: `Creates a new test plan. Provide component IDs as arguments, from a CSV file,
or interactively if no other input is given.`,
	Run: func(cmd *cobra.Command, args []string) {
		s := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
		s.Suffix = " Preparing test plan..."
		s.Start()

		fromCsv, _ := cmd.Flags().GetString("from-csv")
		dependencies, _ := cmd.Flags().GetBool("dependencies")
		creds, _ := cmd.Flags().GetString("creds")

		var componentIds []string
		var err error

		if fromCsv != "" {
			s.Suffix = fmt.Sprintf(" Reading components from %s...", color.CyanString(fromCsv))
			file, err := os.Open(fromCsv)
			if err != nil {
				s.Stop()
				color.Red("\nError: Failed to open file: %v", err)
				os.Exit(1)
			}
			defer file.Close()
			componentIds, err = csv.ParseComponentIdCsv(file)
			if err != nil {
				s.Stop()
				color.Red("\nError: Failed to parse CSV file: %v", err)
				os.Exit(1)
			}
		} else if len(args) > 0 {
			componentIds = args
		} else {
			s.Stop() // Stop for interactive prompt
			componentIds, err = promptForComponentIDs(dependencies)
			if err != nil {
				color.Red("Error during interactive prompt: %v", err)
				os.Exit(1)
			}
			s.Start()
		}

		if len(componentIds) == 0 {
			s.Stop()
			color.Yellow("No component IDs provided. Exiting.")
			return
		}

		discoveryMode := ""
		if dependencies {
			discoveryMode = " and all their dependencies"
		}
		s.Suffix = fmt.Sprintf(" Creating test plan with %d component(s)%s...", len(componentIds), discoveryMode)

		apiClient := client.NewAPIClient(viper.GetString("api_url"))
		planID, err := apiClient.InitiateDiscovery(componentIds, creds, dependencies)
		if err != nil {
			s.Stop()
			color.Red("\nError: Failed to initiate discovery: %v", err)
			os.Exit(1)
		}

		s.Suffix = fmt.Sprintf(" Test plan created (ID: %s). Waiting for component discovery...", color.CyanString(planID))
		finalPlan, err := apiClient.PollForPlanCompletion(planID)
		if err != nil {
			s.Stop()
			color.Red("\nTest plan creation failed.")
			if finalPlan != nil && finalPlan.FailureReason != nil {
				fmt.Printf("Reason: %s\n", *finalPlan.FailureReason)
			} else {
				fmt.Printf("Reason: %v\n", err)
			}
			os.Exit(1)
		}

		s.Stop()
		color.Green("✅ Test plan processing complete!")
		fmt.Printf("\nTest Plan ID: %s\n\n", color.CyanString(planID))
		display.PrintDiscoveryResult(finalPlan)
		color.Yellow("\nTo execute tests, use the 'execute' command with the Plan ID.")
	},
}

func promptForComponentIDs(dependencies bool) ([]string, error) {
	var ids []string
	reader := bufio.NewReader(os.Stdin)
	message := "Enter a Component ID to add to the plan (leave blank to finish):"
	if dependencies {
		message = "Enter a root Component ID to discover dependencies from (leave blank to finish):"
	}

	for {
		fmt.Print(message + " ")
		input, err := reader.ReadString('\n')
		if err != nil {
			return nil, err
		}
		id := strings.TrimSpace(input)
		if id == "" {
			break
		}
		ids = append(ids, id)
	}
	return ids, nil
}

func init() {
	rootCmd.AddCommand(discoverCmd)
	discoverCmd.Flags().String("from-csv", "", "Path to a CSV file with a single column of 'componentId's")
	discoverCmd.Flags().Bool("dependencies", false, "Discover all dependencies for the provided components")
	discoverCmd.Flags().String("creds", "", "The name of the credential profile to use (required)")
	discoverCmd.MarkFlagRequired("creds")
}
